pkg/operation/botanist/component/extensions/dns/dnsprovider.go

```diff
@@ -135,24 +150,39 @@ func (p *provider) Wait(ctx context.Context) error {
 		interval              = 5 * time.Second
 		severeThreshold       = 15 * time.Second
 		timeout               = 2 * time.Minute
+
+		annotationHealthFunc health.Func
 	)
 
-	if err := p.waiter.UntilTimeout(ctx, interval, timeout, func(ctx context.Context) (done bool, err error) {
-		obj := &dnsv1alpha1.DNSProvider{}
-		if err := p.client.Get(
-			ctx,
-			client.ObjectKey{Name: p.values.Name, Namespace: p.namespace},
-			obj,
-		); err != nil {
+	// wait until we see the timestamp annotation, that we set earlier in Deploy, to prevent falsely returning from Wait
+	// in case of stale cache reads
+	if expectedTimestamp, ok := p.dnsProvider.Annotations[v1beta1constants.GardenerTimestamp]; ok {
+		annotationHealthFunc = health.ObjectHasAnnotationWithValue(v1beta1constants.GardenerTimestamp, expectedTimestamp)
+	}
+
+	// TODO: switch to extensions.WaitUntilObjectReadyWithHealthFunction?
```

Comment:
What about those TODOs?

Reply:
I left them for clarification in the PR review here.
Regarding this one: I assume, that there is a reason for not using `extensions.WaitUntilObjectReadyWithHealthFunction` here, probably because of the differences in error code handling/detection in the dns components. But I wasn't sure, but also too reluctant to make the change, as it might have unforeseen consequences. 
Do you think, we can switch to `extensions.WaitUntilObjectReadyWithHealthFunction`? Or better, do you know the reason why it's not used here?

Reply:
Initially, I think the reason for not using `WaitUntilObjectReady` was related to https://github.com/gardener/external-dns-management/issues/161#issuecomment-806486232, i.e., the DNS resources don't implement the `extensionsv1alpha1.Object` interface. `WaitUntilObjectReadyWithHealthFunction` was introduced later and allows to submit custom health functions with special logic for the DNS resources, so I guess it could be used now?